<launch>
  <!-- Arguments -->
  <arg name="map_file" default="$(find dashgo_nav)/maps/2021_0704_LBVS.yaml"/>
  <arg name="keyboard_cmd_vel_topic"    value="cmd_vel"/>
  <arg name="slam_cmd_vel_topic"    value="cmd_vel"/>
  
  <!-- dashgo driver -->
  <include file="$(find dashgo_description)/launch/dashgo_description.launch"/>
  <param name="/use_sim_time" value="false" />
  <node name="arduino" pkg="dashgo_driver" type="dashgo_driver.py" output="screen" respawn="true">
    <rosparam file="$(find dashgo_driver)/config/my_dashgo_params.yaml" command="load" />
  </node>
  
  <!-- nodelet manager -->
  <node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager"/>
  
  <!-- velocity smoother -->
  <arg name="config_file"           value="$(find dashgo_driver)/config/yocs_velocity_smoother.yaml"/>
  <include file="$(find yocs_velocity_smoother)/launch/velocity_smoother.launch">
    <arg name="node_name"             value="velocity_smoother"/>
    <arg name="nodelet_manager_name"  value="nodelet_manager"/>
    <arg name="config_file"           value="$(arg config_file)"/>
    <arg name="raw_cmd_vel_topic"     value="cmd_vel"/>
    <arg name="smooth_cmd_vel_topic"  value="smoother_cmd_vel"/>
    <arg name="robot_cmd_vel_topic"   value="robot_cmd_vel"/>
    <arg name="odom_topic"            value="odom"/>
  </include>

  <!-- move_base -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen" clear_params="true">
    <rosparam file="$(find dashgo_nav)/config/odom/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find dashgo_nav)/config/odom/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find dashgo_nav)/config/odom/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find dashgo_nav)/config/odom/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find dashgo_nav)/config/odom/base_global_planner_param.yaml" command="load" />
    <rosparam file="$(find dashgo_nav)/config/odom/teb_local_planner_params.yaml" command="load" />
    <rosparam file="$(find dashgo_nav)/config/odom/move_base_params.yaml" command="load" />
    <rosparam file="$(find dashgo_nav)/config/odom/prohibition_areas.yaml" command="load" ns="global_costmap/costmap_prohibition_layer" />
    <rosparam file="$(find dashgo_nav)/config/odom/prohibition_areas.yaml" command="load" ns="local_costmap/costmap_prohibition_layer" />
    <remap from="cmd_vel" to="$(arg slam_cmd_vel_topic)"/>
  </node>

  <!-- Map server -->
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)" />

  <!-- AMCL -->
  <include file="$(find dashgo_nav)/launch/include/odom/amcl.launch.xml">
    <arg name="initial_pose_x" value="0.0"/>
    <arg name="initial_pose_y" value="0.0"/>
    <arg name="initial_pose_a" value="0.0"/>
  </include>

  <!-- tf -->
  <node pkg="tf" type="static_transform_publisher" name="base_footprint_to_usb_cam" 
  args="0.17 0 0.39 -1.57 0 -1.57 /base_footprint /usb_cam 40" />
  <node pkg="tf" type="static_transform_publisher" name="base_footprint_to_laser"
   args="0.13 0 0.27 0 0 0 /base_footprint /scan 100" />
  <node pkg="tf" type="static_transform_publisher" name="base_footprint_to_line"
  args="0.13 0 0.27 0 0 0 /base_footprint /laser_line 100" />

  <!-- usb_cam -->
  <node name="usb_cam" pkg="usb_cam" type="usb_cam_node" >
    <param name="video_device" value="/dev/video2" />
    <param name="image_width" value="1280" /><!--640--><!--1280--><!--1920-->
    <param name="image_height" value="720" /><!--480--><!--720--><!--1080-->
    <param name="pixel_format" value="yuyv" />
    <param name="io_method" value="mmap"/>
    <param name="camera_frame_id" value="usb_cam" />
    <!--param name="framerate" value="60"/-->
    <!--param name="autofocus" value="true"/-->
    <param name="camera_info_url" type="string" value="file://$(find usb_cam)/webcam_1280_720.yaml" />
  </node>

  <!-- p+f r2000 -->
  <arg name="frame_id" default="/scan"/>
  <arg name="scanner_ip" default="192.168.1.2"/>
  <arg name="scan_frequency" default="60"/>
  <arg name="samples_per_scan" default="100"/> <!--3600-->
  <node pkg="pepperl_fuchs_r2000" type="r2000_node" name="r2000_node" respawn="true" output="screen">
    <param name="frame_id" value="$(arg frame_id)"/>
    <param name="scanner_ip" value="$(arg scanner_ip)"/>
    <param name="scan_frequency" value="$(arg scan_frequency)"/>
    <param name="samples_per_scan" value="$(arg samples_per_scan)"/>
  </node>

  <!-- laser filter -->
  <node pkg="laser_filters" type="scan_to_scan_filter_chain" name="laser_filter">
    <rosparam command="load" file="$(find laser_filters)/examples/r2000_laser_config.yaml" />
    <remap from="/scan_filtered" to="/scan"/>
  </node>

  <!-- rviz_view -->
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find robot_vision)/config/ar_track_camera.rviz" />

  <!-- teleop_twist_keyboard -->
  <node pkg="teleop_twist_keyboard" type="teleop_twist_keyboard.py" name="teleop_twist_keyboard">
    <remap from="cmd_vel" to="$(arg keyboard_cmd_vel_topic)"/>
  </node>

  <!-- robot_pose_publisher -->
  <node pkg="dashgo_tools" name="robot_pose_publisher" type="robot_pose_publisher">
    <param name="map_frame" type="string" value= '/map' />
    <param name="base_frame" type="string" value= '/base_link' />
    <param name="publish_frequency" type="double" value= '20' />
    <param name="is_stamped" type="bool" value= '0' />
  </node>

  <!-- laser_line_extraction -->
  <node name="line_extractor" pkg="laser_line_extraction" type="line_extraction_node">
    <param name="~frequency" value="10.0" />
    <param name="~frame_id" value="laser_line" />
    <param name="~scan_topic" value="scan" />
    <param name="~publish_markers" value="true" />
    <param name="~bearing_std_dev" value="1e-5" />
    <param name="~range_std_dev" value="0.012" />
    <param name="~least_sq_angle_thresh" value="0.0001" />
    <param name="~least_sq_radius_thresh" value="0.0001" />
    <param name="~max_line_gap" value="0.45" />
    <param name="~min_line_length" value="0.35" />
    <param name="~min_range" value="0.1" />
    <param name="~min_split_dist" value="0.20" />
    <param name="~outlier_dist" value="0.06" />
    <param name="~min_line_points" value="30" />
  </node>

  <!-- map_scoring_node -->
  <node pkg="dashgo_tools" name="map_scoring_node" type="map_scoring_node" output="screen">
    <param name="inputPoseTopic" type="string" value="/robot_pose" />
    <param name="inputMapTopic" type="string" value="/map" />
    <param name="inputScanTopic" type="string" value="/scan" />
    <param name="inputWeightSetTopic" type="string" value="/map_score_weight_set" />
    <param name="minScoreRange1" type="double" value="0.0" />
    <param name="maxScoreRange1" type="double" value="20.0" />
    <param name="scanAngleTolerance" type="double" value="1.0" />
    <param name="scanDistanceTolerance" type="double" value="0.1" />
    <param name="is_outputScan" type="bool" value="1" />
    <rosparam param= "weightSet1" >
    [ 0, 90, 1.0,
      90, 180, 1.0,
      180, 270, 1.0,
      270, 360, 1.0]
    </rosparam>
  </node>

  <!-- pose_update_while_low_score -->
  <node pkg="dashgo_tools" name="pose_update_while_low_score" type="pose_update_while_low_score" output="screen">
    <param name="map_frame" type="string" value= '/map' />
    <param name="base_frame" type="string" value= '/base_link' />
    <param name="inputPoseTopic" type="string" value= '/robot_pose' />
    <param name="inputScoreTopic" type="string" value= '/map_score' />
    <param name="requestTimes" type="int" value= '20' />
    <param name="checkInterval" type="double" value= '1' />
    <param name="updateThresh" type="double" value= '0.05' />
  </node>

  <!-- ar_track_alvar -->
  <node name="ar_track_alvar" pkg="ar_track_alvar" type="individualMarkersNoKinect" respawn="false" >
    <param name="marker_size"           type="double" value="5.0" />
    <param name="max_new_marker_error"  type="double" value="0" /><!--0.08-->
    <param name="max_track_error"       type="double" value="0" /><!--0.2-->
    <param name="output_frame"          type="string" value="/usb_cam" />
    <remap from="camera_image"  to="/usb_cam/image_raw" />
    <remap from="camera_info"   to="/usb_cam/camera_info" />
  </node>

  <!-- artag_update_node -->
  <node pkg="dashgo_tools" name="artag_update_node" type="artag_update_node" output="screen">
    <param name="inputARTagTopic" type="string" value="/visualization_marker" />
    <param name="inputLaserLineTopic" type="string" value="/line_segments" />
    <param name="outputARTagTopic" type="string" value="/visualization_marker_updated" />
    <param name="outputLaserLineTopic" type="string" value="/line_markers_selected" />
    <param name="outputAngleTopic" type="string" value="/marker_angle" />
    <param name="outputUpdatedAngleTopic" type="string" value="/marker_angle_updated" />
    <param name="updateFrequency" type="double" value="10" />
    <!--visualization for selected line-->
    <param name="is_outputLine" type="bool" value="1" />
    <!--visualization for marker angle difference in rqt_plot-->
    <param name="is_outputAngle" type="bool" value="1" />
  </node>

</launch>